types/react/index.d.ts

```diff
@@ -255,8 +255,30 @@ declare namespace React {
         unstable_observedBits?: number;
     }
 
-    type Provider<T> = ComponentType<ProviderProps<T>>;
-    type Consumer<T> = ComponentType<ConsumerProps<T>>;
+    // TODO: similar to how Fragment is actually a symbol, the values returned from createContext,
+    // forwardRef and memo are actually objects that are treated specially by the renderer; see:
+    // https://github.com/facebook/react/blob/v16.6.0/packages/react/src/ReactContext.js#L35-L48
+    // https://github.com/facebook/react/blob/v16.6.0/packages/react/src/forwardRef.js#L42-L45
+    // https://github.com/facebook/react/blob/v16.6.0/packages/react/src/memo.js#L27-L31
+    // However, we have no way of telling the JSX parser that it's a JSX element type or its props other than
+    // by pretending to be a normal component.
+    //
+    // We don't just use ComponentType or SFC types because you are not supposed to attach statics to this
+    // object, but rather to the original function.
+    interface ExoticComponent<P = {}> {
```

Comment:
If we're introducing the concept of exotic components (components without statics, essentially), would it be a good idea to introduce `ExoticComponent` as one of the union types in `React.ComponentType`? To userland I believe`ComponentType` and `ReactType` are all the "all-encompassing" types that can be used as JSX value-based components.

Reply:
Never mind, I just read up on https://github.com/Microsoft/TypeScript/issues/27948.

Reply:
If it's a static added to the exotic component itself, wouldn't it defeat the purpose of it being an exotic component, since that type was created to discourage adding statics to it? I guess it's a smaller subset of the statics allowed in vanilla components.

Reply:
Only forwardRef has that, though, not memo. They're more like subtypes of ExoticComponent.